### PR TITLE
Ensure config consistency on multi apps

### LINKED ISSUE #8: 複数台構成にしたときに設定が反映できないバグがありそう
- 設定をDB化したときに多分そんなことになっていそう
- 変更を加えた際に、DBへの反映 → プロセス内設定キャッシュの更新を行うが、これが多分その実行サーバー1台でしか行われない
- これをちゃんとやるには、設定を毎リクエスト読み込むか、または ...
- 設定に revision を持たせる案

### PR DESCRIPTION
**Pull request by lightnet328**
# Overview
Servers have config cache and those config are not synced.
Therefore, consistency in config can not be ensured on multiple apps environment.

# Changes

Sync config using Redis PubSub when config changed on other server.

# Test

1. Edit `docker-compose.development.yml`

```yml
...
  crowi:
...
    ports:
      - 3000-3001:3000
...
```

2. `docker-compose -f docker-compose.development.yml --scale crowi=2`

3. Browse `localhost:3000` and change config

4. Browse `localhost:3001`

## Related Issues

- Close #8

### CODE REVIEW

**Comment by sotarok** [comment]
I did QA with 3 dynos on Heroku `crowi-dev-pr-435` and I confirmed it worked well.

image

```
2019-05-18T16:57:42.227407+00:00 heroku[web.2]: Starting process with command `node app.js`
2019-05-18T16:57:42.326359+00:00 heroku[web.1]: Starting process with command `node app.js`
2019-05-18T16:57:43.986896+00:00 heroku[web.3]: Starting process with command `node app.js`
2019-05-18T16:57:44.601663+00:00 app[web.2]: 2019-05-18T16:57:44.601Z crowi:crowi PubSubId 618d5f8b-70ce-4276-9b17-a88be3433e92
2019-05-18T16:57:44.802711+00:00 app[web.1]: 2019-05-18T16:57:44.802Z crowi:crowi PubSubId 0e80f94d-2e2b-4b1b-87f6-5593ee96a3e7
2019-05-18T16:57:45.289951+00:00 app[web.2]: 2019-05-18T16:57:45.289Z crowi:crowi Using socket.io-redis
2019-05-18T16:57:45.290394+00:00 app[web.2]: [production] Express server listening on port 14014
2019-05-18T16:57:45.726962+00:00 heroku[web.2]: State changed from starting to up
2019-05-18T16:57:45.678611+00:00 app[web.2]: 2019-05-18T16:57:45.678Z crowi:crowi Config updated by 0e80f94d-2e2b-4b1b-87f6-5593ee96a3e7
2019-05-18T16:57:45.594193+00:00 app[web.1]: 2019-05-18T16:57:45.594Z crowi:crowi Using socket.io-redis
2019-05-18T16:57:45.594795+00:00 app[web.1]: [production] Express server listening on port 45079
2019-05-18T16:57:46.063841+00:00 heroku[web.1]: State changed from starting to up
2019-05-18T16:57:46.913792+00:00 app[web.1]: 2019-05-18T16:57:46.913Z crowi:crowi Websocket CONNECTED, socket.id: uBU_Fba_leyY6sp-AAAA
2019-05-18T16:57:47.102708+00:00 app[web.3]: 2019-05-18T16:57:47.102Z crowi:crowi PubSubId 0939339e-7299-4167-97b1-abbff24fa595
2019-05-18T16:57:48.173123+00:00 app[web.2]: 2019-05-18T16:57:48.173Z crowi:crowi Config updated by 0939339e-7299-4167-97b1-abbff24fa595
2019-05-18T16:57:48.149922+00:00 app[web.3]: 2019-05-18T16:57:48.149Z crowi:crowi Using socket.io-redis
2019-05-18T16:57:48.150542+00:00 app[web.3]: [production] Express server listening on port 24885
2019-05-18T16:57:48.172997+00:00 app[web.1]: 2019-05-18T16:57:48.172Z crowi:crowi Config updated by 0939339e-7299-4167-97b1-abbff24fa595
2019-05-18T16:57:48.458331+00:00 heroku[web.3]: State changed from starting to up
2019-05-18T16:58:06.598790+00:00 heroku[router]: at=info method=GET path="/admin" host=crowi-dev-pr-435.herokuapp.com request_id=a9026e6b-d05c-4782-812d-af125c70d001 fwd="210.224.70.1" dyno=web.2 connect=0msservice=78ms status=304 bytes=315 protocol=https
(snip...)
2019-05-18T16:58:08.036014+00:00 heroku[router]: at=info method=GET path="/admin/app" host=crowi-dev-pr-435.herokuapp.com request_id=9694745d-cf10-47a3-9bee-78ee48bc1e20 fwd="210.224.70.1" dyno=web.1 connect=1ms service=80ms status=200 bytes=26389 protocol=https
(snip...)
2019-05-18T16:58:11.269789+00:00 heroku[router]: at=info method=POST path="/_api/admin/settings/app" host=crowi-dev-pr-435.herokuapp.com request_id=218719df-f107-4def-a171-7d6875e0cbaa fwd="210.224.70.1" dyno=web.2 connect=0ms service=45ms status=200 bytes=410 protocol=https
2019-05-18T16:58:11.254896+00:00 app[web.2]: 2019-05-18T16:58:11.254Z crowi:lib:middlewares req.skipCsrfVerify undefined
2019-05-18T16:58:11.262600+00:00 app[web.2]: 2019-05-18T16:58:11.260Z crowi:routes:admin form content { 'app:title': 'Crowi 3',
2019-05-18T16:58:11.262603+00:00 app[web.2]: 'app:confidential': '',
2019-05-18T16:58:11.262604+00:00 app[web.2]: 'app:fileUpload': '' }
2019-05-18T16:58:11.279644+00:00 app[web.1]: 2019-05-18T16:58:11.279Z crowi:crowi Config updated by 618d5f8b-70ce-4276-9b17-a88be3433e92
2019-05-18T16:58:11.278697+00:00 app[web.3]: 2019-05-18T16:58:11.278Z crowi:crowi Config updated by 618d5f8b-70ce-4276-9b17-a88be3433e92
```

At the starting up the server processes, servers are updated its config cache each other, It is not a bad effect to actual behavior thought.

File: lib/crowi/index.js [hunk 1]
```diff
@@ -107,12 +111,42 @@ Crowi.prototype.isPageId = function(pageId) {
...
+    this.publisher = redis.createClient(this.redisOpts)
+    this.subscriber = redis.createClient(this.redisOpts)
+
+    debug('PubSubId', this.pubSubId)
+
+    if (this.subscriber) {
+      this.subscriber.on('message', (channel, message) => {
+        if (channel !== this.pubSubChannel) return
+        const { pubSubId, config } = JSON.parse(message)
+        if (pubSubId === this.pubSubId) return
```
Comment (sotarok) [inline lib/crowi/index.js]: [memo] If it is the message from itself, ignore.

File: lib/crowi/index.js [hunk 2]
```diff
@@ -107,12 +111,42 @@ Crowi.prototype.isPageId = function(pageId) {
   return false
 }
 
+Crowi.prototype.beforeSetConfig = function(config) {
```
Comment (otofune) [inline lib/crowi/index.js]: [IMO] How about `publishConfig` for name?
I couldn't understand that this function is used for sharing new configuration to another nodes from `beforeSetConfig` name.

File: lib/models/config.js [hunk 3]
```diff
@@ -49,102 +46,100 @@ module.exports = function(crowi) {
...
     Config.count({ ns: 'crowi' }, function(err, count) {
       if (count > 0) {
         return callback(new Error('Application already installed'), null)
       }
-      Config.updateNamespaceByArray('crowi', getArrayForInstalling(), function(err, configs) {
-        Config.updateConfigCache('crowi', configs)
-        return callback(err, configs)
-      })
+      try {
+        return callback(Config.updateConfigByNamespace('crowi', getArrayForInstalling()), null)
```
[the remark this answers is not in the document]
Comment (lightnet328) [reply lib/models/config.js]: Thanks! I didn't notice.
I refactored as fixing a bug

**Pull request merged**